### ISSUE TITLE
cookie date throw error from google https://accounts.google.com/ServiceLogin

### ISSUE THREAD

**Issue by Anderman** [aspnet/HttpAbstractions#507]
I get a cookie from https://accounts.google.com/ServiceLogin with I try to parse with
`Microsoft.Net.Http.Headers.SetCookieHeaderValue.Parse(setCookie)`
But the date format is supported

cookie `GAPS=1:wpZ7lQ_MpSx3RQP9BfRvc7naDmfgkw:4prQKIpNJTstxhRh;Path=/;Expires=Sat, 09-Dec-2017 07:07:03 GMT;Secure`

If I understand [rfc6265 5.1.1.1](http://tools.ietf.org/html/rfc6265#section-5.1.1) correct, all kind of format should be allowed.
But this is not implemented in [HttpRuleParser](https://github.com/aspnet/HttpAbstractions/blob/8ef20c5c5677220dcf7d4e4b6ac14f764d89c92d/src/Microsoft.Net.Http.Headers/HttpRuleParser.cs)

I found a java implemtation of httpdate [here](https://android.googlesource.com/platform/libcore/+/35f9da25ee063a0d0c32f7246b466cff8e57c9ee/luni/src/main/java/libcore/net/http/HttpDate.java)

**Comment by Anderman** [comment]
I create an implementation of RFC-6265, RFC-1123, RFC-850, RFC-5322 for cookie date
That was 3 -10 times faster then the current TryStringToDate.
3 times when the cookiedate is the first dateformat
10 time when cookiedate was the last dateformat

```
  private static readonly Dictionary<string, int> Month = new Dictionary<string, int>
        {
            { "jan",1},
            { "feb",2},
            { "mar",3},
            { "apr",4},
            { "may",5},
            { "jun",6},
            { "jul",7},
            { "aug",8},
            { "sep",9},
            { "oct",10},
            { "nov",11},
            { "dec",12}
        };

        private enum ParseStatus
        {
            InDmy,
            InHms,
            InTz
        }
        public static bool TryStringToDate(string input, out DateTimeOffset result)
        {
            //implementation of RFC-6265, RFC-1123, RFC-850, RFC-5322
            var str = "";
            int MM, yyyy, hh, mm, ss, num, hhTimezone, mmTimezone;
            var dd = MM = yyyy = hh = mm = ss = hhTimezone = -1;
            var i = num = mmTimezone = 0;
            var status = ParseStatus.InDmy;
            var sign = 1;

            foreach (var c in input)
            {
                if (c == 0x20 || c == 0x09 || c > 0x20 && c <= 0x2f || c >= 0x3A && c <= 0x40 || c >= 0x5B && c <= 0x60 || c >= 0x7B && c <= 0x7E)
                {
                    switch (status)
                    {
                        case ParseStatus.InDmy:
                            if (i == 1 || i == 2)
                            {
                                if (mm == -1)
                                {
                                    dd = num;
                                }
                                else
                                {
                                    yyyy = num;
                                    status = ParseStatus.InHms;
                                }
                            }
                            else if (i == 3)
                            {
                                Month.TryGetValue(str.ToLowerInvariant(), out MM);
                            }
                            else if (i == 4)
                            {
                                yyyy = num;
                                status = ParseStatus.InHms;
                            }
                            break;
                        case ParseStatus.InHms:
                            if (hh == -1)
                            {
                                hh = num;
                            }
                            else if (mm == -1)
                            {
                                mm = num;
                            }
                            else
                            {
                                ss = num;
                                status = ParseStatus.InTz;
                            }
                            break;
                        case ParseStatus.InTz:
                            if (c == 0x2c)
                            {
                                sign = -1;
                            }
                            if (hhTimezone == -1)
                                hhTimezone = num;
                            else
                            {
                                mmTimezone = num;
                            }
                            break;
                    }
                    str = "";
                    num = 0;
                    i = 0;
                }
                else if (c >= 0x30 && c <= 0x39)
                {
                    i++;
                    num = num * 10 + (c - 0x30);
                }
                else
                {
                    i++;
                    str = str + c;
                }
            }

            var ts = (hhTimezone == -1) ? new TimeSpan() : new TimeSpan(sign * hhTimezone, mmTimezone, 0);

            if (yyyy >= 1
                && MM >= 1
                && dd >= 1 && dd <= 31
                && hh >= 0 && hh <= 24
                && mm >= 0 && mm <= 60
                && ss >= 0 && ss <= 60
                )
            {
                result = new DateTimeOffset(yyyy, MM, dd, hh, mm, ss, ts);
                return true;
            }
            result = new DateTimeOffset();
            return false;
        }
```

**Comment by glennc** [comment]
The issue here is that the date format in your initial bug isn't supported right? The perf improvement is an aside?

**Comment by Anderman** [comment]
That's correct

**Comment by jkotalik** [comment]
@Tratcher I got a fix for this done, but to avoid this in the future, we should keep dates formats consistent between corefx and our date parsing. Also, I may make an issue for perf (based on what is above) if I get the chance to profile.

**Comment by jkotalik** [comment]
Fixed in #897
